lock-impl/src/main/java/com/palantir/lock/impl/ThreadPooledWrapper.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2017 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.lock.impl;
+
+import java.util.concurrent.Semaphore;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.palantir.common.base.FunctionCheckedException;
+
+public class ThreadPooledWrapper<F> {
+    private static final Logger log = LoggerFactory.getLogger(ThreadPooledWrapper.class);;
+
+    private final Semaphore localThreadPool;
+    private final Semaphore sharedThreadPool;
+    private final F delegate;
+
+    public ThreadPooledWrapper(F delegate, int localThreadPoolSize, Semaphore sharedThreadPool) {
+        this.sharedThreadPool = sharedThreadPool;
+        this.localThreadPool = new Semaphore(localThreadPoolSize);
+        this.delegate = delegate;
+    }
+
+    public F delegate() {
+        return delegate;
+    }
+
+    public <T, K extends Exception> T applyWithPermit(FunctionCheckedException<F, T, K> function) throws K {
+        if (localThreadPool.tryAcquire()) {
+            return applyAndRelease(localThreadPool, function);
+        }
+        if (sharedThreadPool.tryAcquire()) {
+            return applyAndRelease(sharedThreadPool, function);
+        }
+        throw new TooManyRequestsException("ThreadPooledLockService was unable to acquire a permit to assign a server "
```

Comment:
nit: preferably break the line as below instead, to avoid splitting the string:
```
        throw new TooManyRequestsException(
                "ThreadPooledLockService was unable to acquire a permit to assign a server thread to the request.");
```